### DISCUSSION TITLE
Some lat_lon plots have incorrect units

### DISCUSSION THREAD

**Discussion by forsyth2** [E3SM-Project/e3sm_diags#687] [Q&A] [answered]
@evasinha encountered an issue with the units on some plots: https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.eva.sinha/CRP-CTRL/lat_lon/viewer/lat_lon/index.html

TOTSOMC, TOTVEGC, and CPOOL have units displayed as "gC/m^2" and max/min/mean in what appears to be "gC/m^2", but contour lines in what appears to be "kgC/m^2".

```
$ git grep -n TOTSOMC
derivations/acme.py:1861:    "TOTSOMC": OrderedDict(
derivations/acme.py:1862:        [(("TOTSOMC",), lambda v: convert_units(v, target_units="kgC/m^2"))]
driver/default_diags/lat_lon_land_model_vs_model.cfg:857:variables = ["TOTSOMC"]

$ git grep -n TOTVEGC
derivations/acme.py:1858:    "TOTVEGC": OrderedDict(
derivations/acme.py:1859:        [(("TOTVEGC",), lambda v: convert_units(v, target_units="kgC/m^2"))]
driver/default_diags/lat_lon_land_model_vs_model.cfg:887:variables = ["TOTVEGC"]

$ git grep -n CPOOL
derivations/acme.py:1869:    "CPOOL": OrderedDict(
derivations/acme.py:1870:        [(("CPOOL",), lambda v: convert_units(v, target_units="kgC/m^2"))]
driver/default_diags/lat_lon_land_model_vs_model.cfg:15:variables = ["CPOOL"]
```
So, it appears that all three of these variables should be converted to kgC/m^2 yet apparently only the contour lines actually update.

@chengzhuzhang do you know why that might be the case? Thanks!

**Comment by chengzhuzhang** [comment, marked as answer]
@evasinha hi eva, I looked at the configuration file form your run, there is one issue with contour_levels values that I can't reproduce the run. After I corrected the non-monotonic contour-level values. I got the correct results (units + values), [here](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.zhang40/CRP-CTRL/lat_lon/lat_lon/model_vs_model/-TOTSOMC-ANN-global.png). 

The session in .cfg file that is problematic.
[#]
sets = ["lat_lon"]
case_id = "model_vs_model"
variables = ["TOTSOMC"]
seasons = ["ANN", "DJF", "MAM", "JJA", "SON"]
regions = ["global"]
test_colormap = "WhiteBlueGreenYellowRed.rgb"
reference_colormap = "WhiteBlueGreenYellowRed.rgb"
diff_colormap = "BrBG"
contour_levels = [0,**2500.0**,5.0,10.0,15.0,20.0,25.0,30.0,40.0,50.0,60.0,70.0,80.0,100.0]
diff_levels = [-50.0,-40.0,-30.0,-25.0,-20.0,-15.0,-10.0,-5.0,**-2500.0**,0.0,2500.0,5.0,10.0,15.0,20.0,25.0,30.0,40.0,50.0]

What I can't explain is that in my initial effort to reproduce the run with the saved command line at the bottom. The run failed with error: **ValueError: Contour levels must be increasing**.  and no figure is generated. But Eva was able to generate figures, but with wrong units and with re-computed levels...
```
e3sm_diags lat_lon --no_viewer --case_id 'model_vs_model' --sets 'lat_lon' --run_type 'model_vs_model' --variables 'TOTSOMC' --seasons 'ANN' --regions 'global' --regrid_tool 'esmf' --regrid_method 'conservative' --main_title 'TOTSOMC ANN global' --backend 'mpl' --output_format 'png' --canvas_size_w '1212' --canvas_size_h '1628' --figsize '8.5' '11.0' --dpi '150' --arrows --contour_levels '0' '2500.0' '5.0' '10.0' '15.0' '20.0' '25.0' '30.0' '40.0' '50.0' '60.0' '70.0' '80.0' '100.0' --test_name '20230413_CBGCv2.r05.chrysalis.I1850GSWCNPCROP.fnsp' --test_colormap 'WhiteBlueGreenYellowRed.rgb' --reference_name '20230413_CBGCv2.r05.chrysalis.I1850GSWCNPCROP.fnsp' --reference_colormap 'WhiteBlueGreenYellowRed.rgb' --diff_title 'CRP - CTRL' --diff_colormap 'BrBG' --diff_levels '-50.0' '-40.0' '-30.0' '-25.0' '-20.0' '-15.0' '-10.0' '-5.0' '-2500.0' '0.0' '2500.0' '5.0' '10.0' '15.0' '20.0' '25.0' '30.0' '40.0' '50.0' --multiprocessing --num_workers '32' --granulate 'variables' 'seasons' 'plevs' 'regions' --selectors 'sets' 'seasons' --reference_data_path '/lcrc/group/e3sm/ac.eva.sinha/E3SMv2//20230413_CBGCv2.r05.chrysalis.I1850GSWCNPCROP.fnsp/climo' --test_data_path '/lcrc/group/e3sm/ac.eva.sinha/E3SMv2//20230413_CBGCv2.r05.chrysalis.I1850GSWCNPCROP.fnsp/climo' --test_start_yr 501 --test_end_yr 600 --ref_start_yr 501 --ref_end_yr 600 --results_dir '/lcrc/group/e3sm/public_html/diagnostic_output/ac.eva.sinha/CRP-CTRL/lat_lon' 
```
Here is my [TOTVEGC plot](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.zhang40/CRP-CTRL/lat_lon/lat_lon/model_vs_model/-TOTVEGC-ANN-global.png) and [CPOOL plot](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.zhang40/CRP-CTRL/lat_lon/lat_lon/model_vs_model/-CPOOL-ANN-global.png). They look normal. I suspect that the e3sm diags version you have (v2.6.1) may cause part of the issue. I'm on the latest v2.8 version. 

To reproduce for one variable, you could find the command line at the bottom (see figure below), only need to edit the `--result_dir` flag to point to a new path. 

<img width="1214" alt="Screen Shot 2023-05-01 at 4 18 02 PM" src="https://user-images.githubusercontent.com/13056557/235548032-a1c0691d-52fa-47bb-ac6d-6cdffc121750.png">

**Reply by evasinha-pnnl** [reply]
@chengzhuzhang Thank you so much for pointing me to the mistake in TOTSOMC contour levels. I had missed that when modifying the contour levels to kgC/m2 from gC/m2. I was using the default E3SM Diagnostics Package v2.8.0 but since it was still giving me an error (due to the non-monotonic contour-level values), I decided to try an  earlier version of e3sm diags but that was creating incorrect maps.  After correcting the contour level error, I was able to generate the [correct plots](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.eva.sinha/CRP-CTRL/lat_lon/viewer/lat_lon/index.html) using e3sm diags v2.8.

**Reply by chengzhuzhang** [reply]
Thanks for the clarification. Great to know things are working now :)
